src/input/mod.rs

```diff
@@ -0,0 +1,324 @@
+//!     }
+//!
+//!     fn focus_changed(&mut self, seat: &Seat<Self>, focused: Option<&dyn KeyboardHandler<Self>>) {
+//!         // handle focus changes, if you need to ...
+//!     }
+//!     fn cursor_image(&mut self, seat: &Seat<Self>, image: CursorImageStatus) {
+//!         // handle new images for the cursor ...   
+//!     }
+//! }
+//! ```
+//!
+//! ### Run usage
+//!
+//! Once the seat is initialized, you can add capabilities to it.
+//!
+//! Currently, pointer and keyboard capabilities are supported by this module.
+//! [`smithay::wayland::seat`] also provides an abstraction to send touch-events to client,
+//! further helpers are not provided at this point.
+//! [`smithay::wayland::tablet_manager`] also provides client interaction for drawing tablets.
+//!
+//! You can add these capabilities via methods of the [`Seat`] struct:
+//! [`Seat::add_keyboard`] and [`Seat::add_pointer`].
+//! These methods return handles that can be cloned and sent across thread, so you can keep one around
+//! in your event-handling code to forward inputs to your clients.
+//!
+
+use std::sync::{Arc, Mutex};
+
+use crate::utils::user_data::UserDataMap;
+
+pub mod keyboard;
+pub mod pointer;
+
+use self::keyboard::{Error as KeyboardError, KeyboardHandle, KeyboardHandler};
+use self::pointer::{CursorImageStatus, PointerHandle};
+
+/// Handler trait for Seats
+pub trait SeatHandler: Sized {
+    /// [SeatState] getter
+    fn seat_state(&mut self) -> &mut SeatState<Self>;
+
+    /// Callback that will be notified whenever the focus of the seat changes.
+    fn focus_changed(&mut self, seat: &Seat<Self>, focused: Option<&dyn KeyboardHandler<Self>>);
+    /// Callback that will be notified whenever a client requests to set a custom cursor image.
+    fn cursor_image(&mut self, seat: &Seat<Self>, image: CursorImageStatus);
+}
+/// Delegate type for all [Seat] globals.
+///
+/// Events will be forwarded to an instance of the Seat global.
+#[derive(Debug)]
+pub struct SeatState<D: SeatHandler> {
+    pub(crate) seats: Vec<Seat<D>>,
+}
+
+/// A Seat handle
+/// This struct gives you access to the control of the
+/// capabilities of the associated seat.
+///
+/// It is directly inserted in the wayland display by its [`new`](Seat::new) method.
+///
+/// This is an handle to the inner logic, it can be cloned.
+///
+/// See module-level documentation for details of use.
+#[derive(Debug)]
+pub struct Seat<D: SeatHandler> {
+    pub(crate) arc: Arc<SeatRc<D>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct Inner<D> {
+    pub(crate) pointer: Option<PointerHandle<D>>,
+    pub(crate) keyboard: Option<KeyboardHandle<D>>,
+
+    #[cfg(feature = "wayland_frontend")]
+    pub(crate) touch: Option<crate::wayland::seat::TouchHandle>,
+    #[cfg(feature = "wayland_frontend")]
+    pub(crate) global: Option<wayland_server::backend::GlobalId>,
+    #[cfg(feature = "wayland_frontend")]
+    pub(crate) known_seats: Vec<wayland_server::protocol::wl_seat::WlSeat>,
```

Comment:
The problem is not as much about notification, but about storage. I don't a good idea where else to put this, given this is the most logical place. (I suppose we could add a more abstract UserDataMap for that purpose to avoid the circular dependency on the cost of performance.)

Reply:
Coupling between the two modules was further decreased in the last rework, but I don't really see a way to get rid of these fields.